File: backend/app/services/receipt_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException
from datetime import datetime, timezone
from app.models.receipt import Receipt, ReceiptItem, OperationStatus
from app.models.stock_move import MoveType
from app.schemas.receipt import ReceiptCreate, ReceiptUpdate
from app.utils.reference_generator import generate_reference
from app.services.stock_service import create_stock_move


def get_receipts(db: Session, status: str = None):
    query = db.query(Receipt)
    if status:
        query = query.filter(Receipt.status == status)
    return query.order_by(Receipt.created_at.desc()).all()


def get_receipt(db: Session, receipt_id):
    rec = db.query(Receipt).filter(Receipt.id == receipt_id).first()
    if not rec:
        raise HTTPException(status_code=404, detail="Receipt not found")
    return rec


def create_receipt(db: Session, data: ReceiptCreate, user_id):
    ref = generate_reference(db, "REC", Receipt)
    rec = Receipt(
        reference=ref,
        supplier_id=data.supplier_id,
        warehouse_id=data.warehouse_id,
        notes=data.notes,
        created_by=user_id,
        status=OperationStatus.DRAFT,
    )
    db.add(rec)
    db.flush()
    for item in data.items:
        ri = ReceiptItem(
            receipt_id=rec.id,
            product_id=item.product_id,
            location_id=item.location_id,
            quantity=item.quantity,
        )
        db.add(ri)
    db.commit()
    db.refresh(rec)
    return rec


def update_receipt(db: Session, receipt_id, data: ReceiptUpdate):
    rec = get_receipt(db, receipt_id)
    if rec.status != OperationStatus.DRAFT:
        raise HTTPException(status_code=400, detail="Can only edit DRAFT receipts")
    if data.supplier_id is not None:
        rec.supplier_id = data.supplier_id
    if data.notes is not None:
        rec.notes = data.notes
    if data.items is not None:
        db.query(ReceiptItem).filter(ReceiptItem.receipt_id == rec.id).delete()
        for item in data.items:
            ri = ReceiptItem(
                receipt_id=rec.id,
                product_id=item.product_id,
                location_id=item.location_id,
                quantity=item.quantity,
            )
            db.add(ri)
    db.commit()
    db.refresh(rec)
    return rec


def validate_receipt(db: Session, receipt_id):
    rec = get_receipt(db, receipt_id)
    if rec.status != OperationStatus.DRAFT:
        raise HTTPException(status_code=400, detail="Can only validate DRAFT receipts")
    rec.status = OperationStatus.VALIDATED
    rec.validated_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(rec)
    return rec


def confirm_receipt(db: Session, receipt_id, user_id):
    rec = get_receipt(db, receipt_id)
    if rec.status != OperationStatus.VALIDATED:
        raise HTTPException(status_code=400, detail="Can only confirm VALIDATED receipts")
    for item in rec.items:
        create_stock_move(
            db, item.product_id, item.location_id,
            quantity=item.quantity,
            reference=rec.reference,
            move_type=MoveType.RECEIPT,
            source_document_id=rec.id,
            created_by=user_id,
        )
    rec.status = OperationStatus.DONE
    db.commit()
    db.refresh(rec)
    return rec


def cancel_receipt(db: Session, receipt_id):
    rec = get_receipt(db, receipt_id)
    if rec.status == OperationStatus.DONE:
        raise HTTPException(status_code=400, detail="Cannot cancel a completed receipt")
    rec.status = OperationStatus.CANCELLED
    db.commit()
    db.refresh(rec)
    return rec

File: new_frontend/src/services/receiptService.ts
import apiClient from "./apiClient";

export interface ReceiptItemCreate {
  product_id: string;
  location_id: string;
  quantity: number;
}

export interface ReceiptCreate {
  supplier_id?: string;
  warehouse_id: string;
  notes?: string;
  scheduled_date?: string;
  source_document?: string;
  items: ReceiptItemCreate[];
}

export interface ReceiptUpdate {
  supplier_id?: string;
  notes?: string;
  scheduled_date?: string;
  source_document?: string;
  items?: ReceiptItemCreate[];
}

export interface ReceiptItemResponse {
  id: string;
  product_id: string;
  product_name: string | null;
  location_id: string;
  location_name: string | null;
  quantity: number;
  created_at: string;
}

export interface ReceiptResponse {
  id: string;
  reference: string;
  supplier_id: string | null;
  supplier_name: string | null;
  warehouse_id: string;
  warehouse_name: string | null;
  status: string;
  notes: string | null;
  scheduled_date: string | null;
  source_document: string | null;
  created_by: string;
  validated_at: string | null;
  created_at: string;
  updated_at: string;
  items: ReceiptItemResponse[];
}

export const receiptService = {
  getReceipts: (status?: string) =>
    apiClient.get<ReceiptResponse[]>("/receipts", { params: status ? { status } : {} }).then((r) => r.data),

  getReceipt: (id: string) =>
    apiClient.get<ReceiptResponse>(`/receipts/${id}`).then((r) => r.data),

  createReceipt: (data: ReceiptCreate) =>
    apiClient.post<ReceiptResponse>("/receipts", data).then((r) => r.data),

  updateReceipt: (id: string, data: ReceiptUpdate) =>
    apiClient.put<ReceiptResponse>(`/receipts/${id}`, data).then((r) => r.data),

  validateReceipt: (id: string) =>
    apiClient.post<ReceiptResponse>(`/receipts/${id}/validate`).then((r) => r.data),

  confirmReceipt: (id: string) =>
    apiClient.post<ReceiptResponse>(`/receipts/${id}/done`).then((r) => r.data),

  cancelReceipt: (id: string) =>
    apiClient.post<ReceiptResponse>(`/receipts/${id}/cancel`).then((r) => r.data),
};
